File: tv-web/src/pages/sports/dvrFormat.ts
/** Words and numbers for the DVR (port of DvrFormat.kt / RecordingsSections.kt; unit-tested). */
import { DvrState, type DvrJob, type DvrList, type DvrRule, type DvrStorage, type GameRecordingView } from '../../api/tallyDvr';
import { formatRuntime, formatTime } from '../../util/format';

const KB = 1024;

/** `350 MB`, `4.2 GB`, `12 GB`, `1.2 TB` (binary units, as the server counts). */
export function dvrSize(bytes: number): string {
  const gb = bytes / (KB * KB * KB);
  const tb = gb / KB;
  const oneDecimal = (v: number, unit: string): string => (v.toFixed(1) + ' ' + unit).replace('.0 ', ' ');
  if (tb >= 1) return oneDecimal(tb, 'TB');
  if (gb >= 10) return gb.toFixed(0) + ' GB';
  if (gb >= 1) return oneDecimal(gb, 'GB');
  return Math.max(1, bytes / (KB * KB)).toFixed(0) + ' MB';
}

/** `2h 58m`, `42m`: a recording's length. */
export function dvrLength(seconds: number): string {
  return formatRuntime(Math.round(seconds * 10_000_000));
}

const pad2 = (n: number): string => (n < 10 ? '0' : '') + String(n);

/** `42:10`, `1:02:03`: how long a recording has been running. */
export function dvrElapsed(sinceMs: number, nowMs: number): string {
  const total = Math.max(0, Math.floor((nowMs - sinceMs) / 1000));
  const h = Math.floor(total / 3600);
  const m = Math.floor((total % 3600) / 60);
  const s = total % 60;
  return h > 0 ? `${h}:${pad2(m)}:${pad2(s)}` : `${m}:${pad2(s)}`;
}

const MONTHS = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

/** `Sep 24`, month first. */
export function dvrDate(ms: number): string {
  const d = new Date(ms);
  return `${MONTHS[d.getMonth()] ?? ''} ${d.getDate()}`;
}

/** `Sep 24 · 7:12 PM`, or `Today 7:12 PM`. */
export function dvrDayAndTime(ms: number, nowMs: number): string {
  const d = new Date(ms);
  const n = new Date(nowMs);
  const today = d.getFullYear() === n.getFullYear() && d.getMonth() === n.getMonth() && d.getDate() === n.getDate();
  return today ? `Today ${formatTime(d)}` : `${dvrDate(ms)} · ${formatTime(d)}`;
}

const parse = (iso: string | null): number | null => {
  if (iso === null || iso === '') return null;
  const t = Date.parse(iso);
  return isNaN(t) ? null : t;
};

const FREE_SLOT_PREFIX = 'Waiting for a free slot';

/**
 * A job's state in words: "Recording since 7:12 PM", "Waiting for a stream", "Scheduled to record", "Failed: No
 * stream found for this game", "Finishing the recording", "Recorded", "Canceled".
 */
export function recordingStateText(view: Pick<GameRecordingView, 'state' | 'startedAt' | 'reason'>): string {
  switch (view.state) {
    case DvrState.RECORDING: {
      const since = parse(view.startedAt);
      return since !== null ? `Recording since ${formatTime(new Date(since))}` : 'Recording';
    }
    case DvrState.WAITING:
      return view.reason !== null && view.reason.indexOf(FREE_SLOT_PREFIX) === 0 ? view.reason : 'Waiting for a stream';
    case DvrState.SCHEDULED:
      return 'Scheduled to record';
    case DvrState.FINISHING:
      return 'Finishing the recording';
    case DvrState.DONE:
      return 'Recorded';
    case DvrState.FAILED:
      return `Failed: ${view.reason ?? 'The server could not do that.'}`;
    case DvrState.CANCELED:
      return 'Canceled';
    default:
      return view.state;
  }
}

/** "~9 GB · 1.2 TB free on the server" (or only the free space while the estimate is not in). */
export function estimateLine(storage: DvrStorage | null | undefined): string | null {
  if (storage === null || storage === undefined) return null;
  const parts: string[] = [];
  if (storage.estimate !== null && storage.estimate.bytes > 0) parts.push('~' + dvrSize(storage.estimate.bytes));
  if (storage.freeBytes !== null) parts.push(`${dvrSize(storage.freeBytes)} free on the server`);
  return parts.length > 0 ? parts.join(' · ') : null;
}

/** The storage line at the top of RECORDINGS: "1.2 TB free on the server · 14 GB used by recordings". */
export function storageLine(storage: DvrStorage | null): string | null {
  if (storage === null) return null;
  const parts: string[] = [];
  if (storage.freeBytes !== null) parts.push(`${dvrSize(storage.freeBytes)} free on the server`);
  parts.push(`${dvrSize(storage.usedBytes)} used by recordings`);
  return parts.join(' · ');
}

/** "Keep all" / "Keep the last 5". */
export function keepLastText(keepLast: number): string {
  return keepLast <= 0 ? 'Keep all' : `Keep the last ${keepLast}`;
}

/** The choices of the keep-last menu: all, 5, 10, 20. */
export const KEEP_LAST_CHOICES = [0, 5, 10, 20];

const displayName = (t: { shortName: string; abbr: string; name: string }): string => (t.shortName !== '' ? t.shortName : t.abbr !== '' ? t.abbr : t.name);

/** "Otters at Herons" with the board's short names; the job's own title when the teams are unknown. */
export function jobTitle(job: DvrJob): string {
  const a = displayName(job.game.away);
  const h = displayName(job.game.home);
  return a !== '' && h !== '' ? `${a} at ${h}` : job.title;
}

/** The Recordings tab's sections: RECORDING NOW, SCHEDULED, RECORDED, FAILED, TEAM RULES (canceled jobs left out). */
export interface RecordingsSections {
  recordingNow: DvrJob[];
  scheduled: DvrJob[];
  recorded: DvrJob[];
  failed: DvrJob[];
  rules: DvrRule[];
}

export function recordingsSections(list: DvrList): RecordingsSections {
  const scheduled = list.jobs.filter((j) => j.state === DvrState.SCHEDULED || j.state === DvrState.WAITING);
  return {
    recordingNow: list.jobs.filter((j) => j.state === DvrState.RECORDING || j.state === DvrState.FINISHING),
    scheduled: scheduled
      .map((j, i) => ({ j, i }))
      .sort((a, b) => (a.j.game.start < b.j.game.start ? -1 : a.j.game.start > b.j.game.start ? 1 : a.i - b.i))
      .map((x) => x.j),
    recorded: list.jobs.filter((j) => j.state === DvrState.DONE),
    failed: list.jobs.filter((j) => j.state === DvrState.FAILED),
    rules: list.rules.filter((r) => r.kind === 'team'),
  };
}

export const sectionsEmpty = (s: RecordingsSections): boolean =>
  s.recordingNow.length === 0 && s.scheduled.length === 0 && s.recorded.length === 0 && s.failed.length === 0 && s.rules.length === 0;

/** A recording in progress: "MLB · 42:10 · 1.3 GB" (finishing: "MLB · Finishing the recording"). */
export function recordingNowMeta(job: DvrJob, nowMs: number): string {
  const parts = [job.game.league];
  if (job.state === DvrState.FINISHING) parts.push('Finishing the recording');
  else {
    const started = parse(job.startedAt);
    if (started !== null) parts.push(dvrElapsed(started, nowMs));
    parts.push(dvrSize(job.bytes));
  }
  return parts.filter((p) => p !== '').join(' · ');
}

/** A scheduled job: "MLB · Today 7:05 PM · Every Otters game" (or "Waiting for a stream" once it is due). */
export function scheduledMeta(job: DvrJob, rules: readonly DvrRule[], nowMs: number): string {
  const parts = [job.game.league];
  if (job.state === DvrState.WAITING) parts.push(recordingStateText({ state: job.state, startedAt: null, reason: job.reason }));
  else {
    const start = parse(job.game.start);
    if (start !== null) parts.push(dvrDayAndTime(start, nowMs));
  }
  const rule = rules.find((r) => r.id === job.ruleId && r.kind === 'team');
  if (rule !== undefined) parts.push(rule.title);
  return parts.filter((p) => p !== '').join(' · ');
}

/** A finished recording's card line: "MLB · Sep 24 · 2h 58m". */
export function recordedMeta(job: DvrJob): string {
  const parts: string[] = [];
  if (job.game.league !== '') parts.push(job.game.league);
  const start = parse(job.game.start);
  if (start !== null) parts.push(dvrDate(start));
  if (job.seconds > 0) parts.push(dvrLength(job.seconds));
  return parts.join(' · ');
}

/** A team rule: "MLB · Keep the last 5" (any league: "All leagues · Keep all"). */
export function ruleMeta(rule: DvrRule, list: DvrList): string {
  const job = list.jobs.find((j) => j.game.leaguePath.toLowerCase() === rule.leaguePath.toLowerCase() && rule.leaguePath !== '');
  const tail = rule.leaguePath.split('/').pop() ?? '';
  const league = job !== undefined ? job.game.league : tail !== '' ? tail.toUpperCase() : 'All leagues';
  return `${league} · ${keepLastText(rule.keepLast)}`;
}

File: app/src/main/java/io/github/scdouglas1999/tally/dvr/DvrFormat.kt
package io.github.scdouglas1999.tally.dvr

import android.text.format.DateFormat
import androidx.compose.runtime.Composable
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import com.github.damontecres.wholphin.R
import io.github.scdouglas1999.tally.media.kit.formatRuntime
import java.time.DateTimeException
import java.time.Duration
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

/** Words and numbers for the DVR: sizes, times, a job's state. */
object DvrFormat {
    private const val KB = 1024.0
    private const val TICKS_PER_SECOND = 10_000_000L

    /** `350 MB`, `4.2 GB`, `12 GB`, `1.2 TB` (binary units, as the server counts). */
    fun size(bytes: Long): String {
        val gb = bytes / (KB * KB * KB)
        val tb = gb / KB
        return when {
            tb >= 1 -> "%.1f TB".format(Locale.US, tb).replace(".0 ", " ")
            gb >= 10 -> "%.0f GB".format(Locale.US, gb)
            gb >= 1 -> "%.1f GB".format(Locale.US, gb).replace(".0 ", " ")
            else -> "%.0f MB".format(Locale.US, (bytes / (KB * KB)).coerceAtLeast(1.0))
        }
    }

    /** `2h 58m`, `42m`: a recording's length. */
    fun length(seconds: Double): String = formatRuntime((seconds * TICKS_PER_SECOND).toLong())

    fun instant(iso: String?): Instant? =
        try {
            iso?.let { OffsetDateTime.parse(it).toInstant() }
        } catch (_: DateTimeException) {
            null
        }

    /** `42:10`, `1:02:03`: how long a recording has been running. */
    fun elapsed(
        since: Instant,
        now: Instant = Instant.now(),
    ): String {
        val total = Duration.between(since, now).seconds.coerceAtLeast(0)
        val h = total / 3600
        val m = (total % 3600) / 60
        val s = total % 60
        return if (h > 0) "%d:%02d:%02d".format(Locale.US, h, m, s) else "%d:%02d".format(Locale.US, m, s)
    }

    /** `Sep 24`, month first. */
    fun date(instant: Instant): String {
        val pattern = DateFormat.getBestDateTimePattern(Locale.US, "MMMd")
        return DateTimeFormatter.ofPattern(pattern, Locale.US).format(instant.atZone(ZoneId.systemDefault()))
    }

    /** `7:12 PM` (or `19:12` on a 24-hour device). */
    @Composable
    fun time(instant: Instant): String {
        val is24 = DateFormat.is24HourFormat(LocalContext.current)
        val pattern = if (is24) "H:mm" else "h:mm a"
        return DateTimeFormatter.ofPattern(pattern, Locale.US).format(instant.atZone(ZoneId.systemDefault()))
    }

    /** `Sep 24 · 7:12 PM`, or only the time for today. */
    @Composable
    fun dayAndTime(instant: Instant): String {
        val zone = ZoneId.systemDefault()
        val today = Instant.now().atZone(zone).toLocalDate()
        return if (instant.atZone(zone).toLocalDate() == today) {
            stringResource(R.string.tally_dvr_today_at, time(instant))
        } else {
            date(instant) + " · " + time(instant)
        }
    }
}

/**
 * A job's state in words: "Recording since 7:12 PM", "Waiting for a stream", "Scheduled to record", "Failed: No
 * stream found for this game", "Finishing the recording", "Recorded", "Canceled".
 */
@Composable
fun recordingStateText(view: GameRecordingView): String =
    when (view.state) {
        DvrState.RECORDING -> {
            DvrFormat.instant(view.startedAt)?.let { stringResource(R.string.tally_dvr_state_recording_since, DvrFormat.time(it)) }
                ?: stringResource(R.string.tally_dvr_state_recording)
        }

        DvrState.WAITING -> {
            // The server says why: no channel carries the game yet, or every recording slot is taken.
            view.reason?.takeIf { it.startsWith(FREE_SLOT_PREFIX) } ?: stringResource(R.string.tally_dvr_state_waiting)
        }

        DvrState.SCHEDULED -> {
            stringResource(R.string.tally_dvr_state_scheduled)
        }

        DvrState.FINISHING -> {
            stringResource(R.string.tally_dvr_state_finishing)
        }

        DvrState.DONE -> {
            stringResource(R.string.tally_dvr_state_done)
        }

        DvrState.FAILED -> {
            stringResource(R.string.tally_dvr_state_failed, view.reason ?: stringResource(R.string.tally_dvr_error_generic))
        }

        DvrState.CANCELED -> {
            stringResource(R.string.tally_dvr_state_canceled)
        }

        else -> {
            view.state
        }
    }

private const val FREE_SLOT_PREFIX = "Waiting for a free slot"

/** "Keep all" / "Keep the last 5". */
@Composable
fun keepLastText(keepLast: Int): String =
    if (keepLast <= 0) {
        stringResource(R.string.tally_dvr_keep_all)
    } else {
        stringResource(R.string.tally_dvr_keep_last, keepLast)
    }

/** The choices of the keep-last sheet: all, 5, 10, 20. */
val KeepLastChoices = listOf(0, 5, 10, 20)
